state/unit_ops.go

```diff
@@ -189,7 +189,7 @@ func (op *unitSetStateOperation) fields(currentDoc unitStateDoc) (bson.D, bson.D
 		if len(rState) == 0 {
 			unsetFields = append(unsetFields, bson.DocElem{Name: "relation-state"})
 		} else if matches := currentDoc.relationStateMatches(rState); !matches {
-			setFields = append(setFields, bson.DocElem{"relation-state", rState})
+			setFields = append(setFields, bson.DocElem{"relation-state", mergeRelations(currentDoc.RelationState, rState)})
```

Comment:
You need to pull mergeRelations result into a var and pass that to the Check call below instead of `rState`

Reply:
Why did we include the uniter internal state data in the quota? Isn't there a potential to stop the unit from working?

Reply:
make the change

Reply:
As per #11390 we have two quotas, one for the charm state and one for the uniter state (uniter, relations, storage, meter status).  The code in this method uses separate quota checkers for applying the quota limits. Because `SetState` is not called at the same time for *both* charm and uniter state, charm quota checks can independently fail without affecting the storage of the uniter's state.